Drivers/Src/Display.c

```diff
@@ -1,325 +1,280 @@
+	OSMutexPost(&DisplayQ_Mutex, OS_OPT_POST_NONE, &err);
+	assertOSError(OS_DISPLAY_LOC, err);
 
-    char number[12]; // To store converted int
-    sprintf(number, "%ld", val);
+	if(success){
+		OSSemPost(&DisplayQ_Sem4, OS_OPT_POST_NONE, &err);
+		assertOSError(OS_DISPLAY_LOC, err);
+	}
 
-    BSP_UART_Write(DISP_OUT, number, strlen(number));
-    BSP_UART_Write(DISP_OUT, (char *)TERMINATOR, strlen(TERMINATOR));
+	return success ? SUCCESS : ERROR;
+}
 
-    // Get a response from the display
-    char buf[8];
-    BSP_UART_Read(DISP_OUT, buf);
-    return (IsNextionFailure(buf[0])) ? ERROR : SUCCESS;
+ErrorStatus Display_SendNext(){
+	Display_Cmd_t cmd;
+
+	OS_ERR err;
+	CPU_TS ticks;
+    
+	OSSemPend(&DisplayQ_Sem4, 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+	assertOSError(OS_DISPLAY_LOC, err);
+	
+	OSMutexPend(&DisplayQ_Mutex, 0, OS_OPT_POST_NONE, &ticks, &err);
+	assertOSError(OS_DISPLAY_LOC, err);
+	
+	bool result = disp_fifo_get(&msg_queue, &cmd);
+	OSMutexPost(&DisplayQ_Mutex, OS_OPT_POST_NONE, &err);
+	assertOSError(OS_SEND_CAN_LOC, err);
+	
+	if(result == ERROR)
+		return ERROR;
+	
+	char msg[32];
+	if((cmd.numArgs == 1 && cmd.args != NULL && cmd.argTypes != NULL) && cmd.op != NULL && cmd.attr != NULL){	// Assignment commands have only 1 arg, an operator, and an attribute
+		if(cmd.argTypes[0]){
+			sprintf(msg, "%s.%s%s%d", cmd.compOrCmd, cmd.attr, cmd.op, cmd.args[0].num);
+		}
+		else{
+			sprintf(msg, "%s.%s%s%s", cmd.compOrCmd, cmd.attr, cmd.op, cmd.args[0].str);
+		}
+	}
+	else if(cmd.op == NULL && cmd.attr == NULL){	// Operational commands have no attribute and no operator, just a command and >= 0 arguments
+		sprintf(msg, "%s", cmd.compOrCmd);
+		if(cmd.numArgs == 1 && cmd.argTypes != NULL && cmd.args != NULL){	// If there are arguments
+			strcat(msg, " ");
+			for(int i=0; i<cmd.numArgs; i++){
+				char arg[16];
+				if(cmd.argTypes[i]){
+					sprintf(arg, "%d", cmd.args[i].num);
+				}
+				else{
+					sprintf(arg, "%s", cmd.args[i].str);
+				}
+
+				strcat(msg, arg);
+
+				if(i<cmd.numArgs-1){	// delimiter
+					strcat(msg, ", ");
+				}
+			}
+		}
+	}
+	else{	// Invalid command
+		return ERROR;
+	}
+	
+	BSP_UART_Write(DISP_OUT, msg, strlen(msg));
+	BSP_UART_Write(DISP_OUT, (char*)TERMINATOR, strlen(TERMINATOR));
+
+	char buf[8];
+	BSP_UART_Read(DISP_OUT, buf);
+	if(buf[0] == 0x01){	// Successful command
+		return SUCCESS;
+	}
+	return ERROR;
 }
 
-/*
- * Sets a component's visiblity
- */
-ErrorStatus setComponentVisibility(enum CommandString_t comp, bool vis)
-{
-    char out[24];
-    sprintf(out, "%s %s,%d%s", CommandStrings[VIS], CommandStrings[comp], vis ? 1 : 0, TERMINATOR);
+ErrorStatus Display_SetComponent(Component_t comp, uint8_t val){
+	// For components that are on/off
+	if(comp <= MOTOR && val <= 1){
+		// If blinkers, set the blink toggle of the components instead of the visibility
+		if(comp == LEFT || comp == RIGHT || comp == HZD){
+			Display_Cmd_t toggleCmd = {(char*)compStrings[comp],"en","=",1,{true},
```

Comment:
Using a designated initializer might make this a bit clearer.

Reply:
Right, I did that for the union but you're saying for the struct as a whole? Sounds like a good idea to me